source/stun_serializer.c

```diff
@@ -0,0 +1,862 @@
+             * attribute. */
+            result = STUN_RESULT_INVALID_ATTRIBUTE_ORDER;
+        }
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        /* Update flags. */
+        if( attributeType == STUN_ATTRIBUTE_TYPE_FINGERPRINT )
+        {
+            pCtx->flags |= STUN_FLAG_FINGERPRINT_ATTRIBUTE;
+        }
+        if( attributeType == STUN_ATTRIBUTE_TYPE_MESSAGE_INTEGRITY )
+        {
+            pCtx->flags |= STUN_FLAG_INTEGRITY_ATTRIBUTE;
+        }
+
+        if( pCtx->pStart != NULL )
+        {
+            /* Write Attribute type, length and value. */
+            WRITE_UINT16( (uint8_t *) &( pCtx->pStart[ pCtx->currentIndex ] ),
+                        attributeType );
+
+            WRITE_UINT16( (uint8_t *) &( pCtx->pStart[ pCtx->currentIndex + STUN_ATTRIBUTE_HEADER_LENGTH_OFFSET ] ),
+                        attributeValueLength );
+
+            WRITE_UINT64( (uint8_t *) &( pCtx->pStart[ pCtx->currentIndex + STUN_ATTRIBUTE_HEADER_VALUE_OFFSET ] ),
+                        attributeValue );
+        }
+
+        pCtx->currentIndex += STUN_ATTRIBUTE_TOTAL_LENGTH( attributeValueLength );
+    }
+
+    return result;
+}
+/*-----------------------------------------------------------*/
+
+StunResult_t StunSerializer_Init( StunContext_t * pCtx,
+                                  uint8_t * pBuffer,
+                                  size_t bufferLength,
+                                  const StunHeader_t * pHeader )
+{
+    StunResult_t result = STUN_RESULT_OK;
+
+    init_endianness();
+
+    if( ( pCtx == NULL ) ||
+        ( pHeader == NULL ) ||
+        ( ( pBuffer != NULL ) &&
+         ( bufferLength < STUN_HEADER_LENGTH ) ) )
+    {
+        result = STUN_RESULT_BAD_PARAM;
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        pCtx->pStart = pBuffer;
+        pCtx->totalLength = bufferLength;
+        pCtx->currentIndex = 0;
+        pCtx->flags = 0;
+
+        if ( pCtx->pStart != NULL )
+        {
+            WRITE_UINT16( (uint8_t *) &( pCtx->pStart[ pCtx->currentIndex ] ),
+                      pHeader->messageType );
+
+            /* Message length is updated in finalize. */
+            WRITE_UINT16( (uint8_t *) &( pCtx->pStart[ pCtx->currentIndex + STUN_HEADER_MESSAGE_LENGTH_OFFSET ] ),
+                        0 );
+
+            WRITE_UINT32( (uint8_t *) &( pCtx->pStart[ pCtx->currentIndex + STUN_HEADER_MAGIC_COOKIE_OFFSET ] ),
+                        STUN_HEADER_MAGIC_COOKIE );
+
+            memcpy( ( void * ) &( pCtx->pStart[ pCtx->currentIndex + STUN_HEADER_TRANSACTION_ID_OFFSET ] ),
+                    &( pHeader->transactionId[ 0 ] ),
+                    STUN_HEADER_TRANSACTION_ID_LENGTH );
+        }
+
+        pCtx->currentIndex += STUN_HEADER_LENGTH;
+    }
+
+    return result;
+}
+/*-----------------------------------------------------------*/
+
+StunResult_t StunSerializer_xorIpAddress( StunAttributeAddress_t *ipAddress,
```

Comment:
```suggestion
StunResult_t StunSerializer_XorIpAddress( StunAttributeAddress_t *ipAddress,
```
Should xor be Xor as coding convention?